### DISCUSSION TITLE
[Critical] Search API returns wrong results for pull requests

### DISCUSSION THREAD

**Discussion by bselwe** [community/community#53799] [Apps, API and Webhooks]
### Select Topic Area

Bug

### Body

The search API stopped working correctly today. It does not return all results when using the `draft` qualifier, see example below:

| Query: `is:pr is:open` (returns all open PRs correctly, 10 results) |
|---|
| ![CleanShot 2023-04-26 at 17 13 13](https://user-images.githubusercontent.com/26686598/234621142-0a131fa5-3211-4268-9a62-85d7b5021dd7.png) |

| Query: `is:pr is:open draft:false` (returns only 1 result, should be 8 results as you can see from the list above) |
|---|
![CleanShot 2023-04-26 at 17 11 47](https://user-images.githubusercontent.com/26686598/234620722-6367886e-4fb4-45eb-8810-c5bf2ceead1d.png)

It does not seem to happen in all repositories, but we've managed to reproduce this issue in one of our private repositories: `joinbuzz/buzz`.

We've noticed the same inconsistency using [global search API](https://docs.github.com/en/search-github/searching-on-github/searching-issues-and-pull-requests). This is critical for us as we're heavily dependent on this API and it produces unexpected results in the last days.

**Comment by SarthkDeshmukh** [comment]
The core issue: GitHub's Search API runs on a separate index from the main REST API. That index is not real-time — it has a lag, sometimes minutes, sometimes longer during high-traffic periods. So when you create or update a PR and immediately search for it, you're querying stale data. The PR exists, the main API knows about it, but Search hasn't indexed it yet.
This means GET /search/issues?q=is:pr and GET /repos/{owner}/{repo}/pulls can return genuinely different results for the same repository at the same point in time. It's not a bug in your code — it's an architectural reality of how GitHub's search works.
Specific things that trip people up:
- Newly opened PRs not appearing for 30-60 seconds after creation
- Recently merged PRs still showing as open in search results
- Label or assignee filters returning stale matches
- Repo forks sometimes indexed under wrong parent
What I do instead:

For anything requiring accuracy — use GET /repos/{owner}/{repo}/pulls with filters, not the Search API
Use Search only for discovery across repos where slight staleness is acceptable
If you absolutely need Search, build in a retry with exponential backoff and accept eventual consistency

One thing worth knowing: the Search API also has a stricter rate limit — 30 requests/minute authenticated vs 10 unauthenticated — completely separate from the REST API's limits. So if you're hitting both in the same app, track them independently.

Treat GitHub Search like a search engine, not a database. It's fast and powerful for discovery, but eventual consistency is the contract, not an exception.
